BEIMA.Backend/DeviceFunctions/UpdateDevice.cs

```diff
@@ -40,60 +43,109 @@ public static class UpdateDevice
                 return new BadRequestObjectResult(Resources.InvalidIdMessage);
             }
 
-            Device device;
+            // Parse out necessary data            
             var mongo = MongoDefinition.MongoInstance;
+            IFormCollection reqForm;
+            UpdateDeviceRequest data;
+            Device device;
+            DeviceType deviceType;
+            ObjectId deviceTypeId;
+            ObjectId? buildingId;
             try
             {
-                string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
-                var data = JsonConvert.DeserializeObject<UpdateDeviceRequest>(requestBody);
-                device = new Device(
-                     ObjectId.Parse(id),
-                     ObjectId.Parse(data.DeviceTypeId),
-                     data.DeviceTag,
-                     data.Manufacturer,
-                     data.ModelNum,
-                     data.SerialNum,
-                     data.YearManufactured,
-                     data.Notes
-                 );
+                reqForm = await req.ReadFormAsync();
+                data = JsonConvert.DeserializeObject<UpdateDeviceRequest>(reqForm["data"]);                              
+                
+                var deviceDocument = mongo.GetDevice(new ObjectId(id));
+                if (deviceDocument == null)
+                {
+                    return new NotFoundObjectResult(Resources.DeviceNotFoundMessage);
+                }
+
+                device = BsonSerializer.Deserialize<Device>(deviceDocument);
+                deviceTypeId = ObjectId.Parse(data.DeviceTypeId);
+
+                var deviceTypeDocument = mongo.GetDeviceType(deviceTypeId);
+                deviceType = BsonSerializer.Deserialize<DeviceType>(deviceTypeDocument);
 
                 var reqBuildingId = data.Location.BuildingId;
-                ObjectId? buildingId = reqBuildingId != null ? ObjectId.Parse(reqBuildingId) : null;
+                buildingId = reqBuildingId != null ? ObjectId.Parse(reqBuildingId) : null;
+            }
+            catch (Exception)
+            {
+                return new BadRequestObjectResult(Resources.CouldNotParseBody);
+            }
 
-                device.SetLocation(
-                    buildingId,
-                    data.Location.Notes,
-                    data.Location.Latitude,
-                    data.Location.Longitude
-                );
+            if (!ObjectId.TryParse(data.DeviceTypeId, out _))
+            {
+                return new BadRequestObjectResult(Resources.InvalidIdMessage);
+            }            
+
+            // Set Data to new values
+            device.DeviceTypeId = deviceTypeId;
```

Comment:
What is the reasoning for moving this out here/not using the constructor?

Reply:
It was mostly I wanted to move all of the parsing objectId logic into the trycatch, but still keeps settings all the values in one spot. That and if the parsing failed I felt that the couldnotparsebody was a fitting error message.

Reply:
But this is outside of the trycatch though right?

Reply:
device type id gets assigned inside the trycatch at line 66

Reply:
Yeah, but isn't it safer to include all of these device assignments inside the trycatch? Just so that in the off chance something does actually go wrong here, we still return a could not parse message.

Reply:
Fair enough. Moved it back inside